torch/distributed/elastic/utils/store.py

```diff
@@ -29,17 +49,20 @@ def get_all(store, rank: int, prefix: str, size: int):
      value3 = values[2] # retrieves the data for key torchelastic/data2
 
     """
-    data_arr = []
-    for idx in range(size):
-        data = store.get(f"{prefix}{idx}")
-        data_arr.append(data)
-    store.set(f"{prefix}{rank}.FIN", b"FIN")
+    data_arr = store.multi_get(
+        [f"{prefix}{idx}" for idx in range(world_size)]
+    )
+
+    barrier_key = _barrier_nonblocking(
+        store=store,
+        world_size=world_size,
+        key_prefix=f"{prefix}/finished",
+    )
     if rank == 0:
         # Rank0 runs the TCPStore daemon, as a result it needs to exit last.
         # Otherwise, the barrier may timeout if rank0 process finished the work
         # before other processes finished `get_all` method
-        for node_rank in range(size):
-            store.get(f"{prefix}{node_rank}.FIN")
+        store.get(barrier_key)
```

Comment:
I think we have a 10min timeout for this get? In the case other ranks hang for some reason, then this will exit? I haven't checked but as long as we propagate the error, this might be fine.

Reply:
yes, there's a timeout on this -- defaults to 300s or 5 minutes